accepted/0004-undo-redo.md

```diff
@@ -0,0 +1,73 @@
+- Start Date: 2020-01-03
+- RFC PR: #5
+- Mantis Issue: N/A
+
+# Summary
+Create a undo system that is capable of recording the state of OBS actions to
+easily revert the state forwards or backwards in time.
+
+# Motivation
+When working with large and intricate processes - whether it be scenes, sources,
+etc. - accidents are bound to happen. Currently, there is no way to undo
+committed actions. As a result, work can easily be lost at the expense of hours
+of effort.
+
+Implementing a proper undo/redo system would be a drastic increase in
+quality of life for the majority of users.
+
+# Design
+## UX
+The UX would model that of most other applications. Under edit, there would be 2
+options: 'Undo' and 'Redo'. When an action is comitted, for instance deleting a
+scene, the undo button becomes activated, allowing the user to select it.
+
+When activated, it undos the action and then activates the Redo button, which
+allows the user to essentially 'undo' their previous undo. If the user makes
+multiple undos, then the users would also have multiple redos available.
+
+When a user undos and then commits another action, the redos are no longer
+possible as the tree has diverged into a new path.
+
+Also, like many other applications, the undo/redo keys should be binded to Ctrl-Z
+and Ctrl-Y
+
+## Functionality
+### Core Functionality
+The undo/redo system work by saving states. However, saving the entire state of
+OBS would use up more memory than necessary when a small action is committed. In
+order to prevent this, only the state of the object the action was committed on
+would be saved, alongside the action committed. This will allow more seamless
+integration into OBS without disturbing the flow, and also keep the amount of
+memory low. 
+
+Note, the same methodology would work for the redo system.
+
+The method of saving states would be in combination with 2 linked lists that
+stores the header for redo and undo states. I chose linked lists because it
+allows O(1) insertion that allows for the easy tracking of the most recent undo
+action.
+
+By using smart pointers in the linked list abstraction, when the undo action is
+done and moves onto the next undo state, the state object can be more easily
+discarded. This is more helpful with the redo state because there could a large
+redo list and when discared, memory will automatically be freed reducing memory
+leaks.
+
+### Memory
+The current design of only saving the state of the local area is intended to
+reduce memory consumption. However, over the course of a long lifetime, lots of
+memory would eventually be used by having a long tree. After a certain length n 
+(perhaps customizable?) the end of the chain should be saved to disk and removed
+on close.
+
+Having disk read/write operations for each action would be costly. In
+order to reduce that load it should only save/read in sets of n when
+it needs to.
+
+### Scope
```

Comment:
This scope needs to be significantly more specific for each of the items.

Scenes:
* Add
* Rename
* Reorder
* Delete

Sources/Scene Items:
* Add
* Properties (detailed scope here too, undo individual properties, or only Save/Cancel?)
* Rename
* Transform
  * Mouse drag: on mouse release
  * Keyboard: should this work on a timeframe where the keyboard moving the element should stop for at least 1 second to 'save' the undo
  * Transform submenu, each individual menu action trigger
  * Edit Transform dialog. Currently this sort-of saves a mini-undo-stack thanks to Qt, we want to keep this functionality, but allow bigger undos on Save/Cancel
* Reorder
  * Do we store how many rows it moved up/down, or where it was placed in relation to other sceneitems?
* Move into group
  * Keep in mind this also can do strange things to transform information, and it would likely be a good idea to store before/after of transform
* Delete

Filters:
* Add
* Properties (when the dialog is opened/closed, or when the user switches to the filter in the list?)
* Rename
* Reorder (if the attempt at adding drag & drop support is any indicator, this might get compicated)
* Delete

Audio Controls:
* Are we talking about just the Input Mixer, or Advanced Audio Properties too?
* I personally wouldn't see the need of undoing mute state or monitoring state
* Volume is a must

Transitions:
* Does this include creating/adding Quick Transitions? I vote that come later
* Add
* Properties (same question as Sources)
* Delete

Reply:
> Properties (detailed scope here too, undo individual properties, or only Save/Cancel?)

I would say just Save/Cancel.

> Keyboard: should this work on a timeframe where the keyboard moving the element should stop for at least 1 second to 'save' the undo

Sounds about a good default! I think this might be an option for General Settings

> Move into group. Keep in mind this also can do strange things to transform information, and it would likely be a good idea to store before/after of transform

Sounds tricky

> Audio Controls:

I agree, probably just Volume is needed. All other controls we already have fine control

Reply:
> Reorder

I think most times a reorder is to move an element specifically in-front of/behind other specific elements, so perhaps object-related like as in a linked list may be most appropriate, although that may be more complex than needed.

> Delete

Just to clarify, this will need to hold as much info as possible on the deleted object.